### PR TITLE
[Backport 2.28] Fix fake cases listed of compat.sh

### PR DESCRIPTION
**Pull request by mpg**
## Description

This is the 2.28 backport of #8586.

The backport of the 1st commit is not trivial as the loops are nested differently in 2.28 (the various `add_xxx_ciphersuites` function access the `MODE` variable).

The backport of the 2nd commit is trivial.

## PR checklist

- [x] **changelog** not required, testing only
- [x] **backport** this is the backport
- [x] **tests** not required - manually checked the output of running the script without options, and with `--list-test-cases` before and after, and checking that all differences are the ones we expect.

### CODE REVIEW

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

File: tests/compat.sh [hunk 1]
```diff
@@ -278,12 +283,6 @@ filter_ciphersuites()
         # Ciphersuite for GnuTLS
         G_CIPHERS=$( filter "$G_CIPHERS" )
     fi
-
-    # For GnuTLS client -> Mbed TLS server,
-    # we need to force IPv4 by connecting to 127.0.0.1 but then auth fails
-    if is_dtls "$MODE" && [ "X$VERIFY" = "XYES" ]; then
-        G_CIPHERS=""
-    fi
 }
```
Comment (gilles-peskine-arm) [inline tests/compat.sh:287]: 2.28 only wrinkle (because it's about protocol versions before 1.2):

```
            # TLS-RSA-WITH-NULL-SHA256 is a (D)TLS 1.2-only cipher suite,
            # like all SHA256 cipher suites. But Mbed TLS supports it with
            # (D)TLS 1.0 and 1.1 as well. So do ancient versions of GnuTLS,
            # but this was considered a bug which was fixed in GnuTLS 3.4.7.
            # Check the GnuTLS support list to see what the protocol version
            # requirement is for that cipher suite.
            if [ `minor_ver "$MODE"` -ge 3 ] || {
                   [ `minor_ver "$MODE"` -gt 0 ] &&
                   $GNUTLS_CLI --list | grep -q '^TLS_RSA_NULL_SHA256.*0$'
               }
            then
                M_CIPHERS="$M_CIPHERS                           \
                    TLS-RSA-WITH-NULL-SHA256                    \
                    "
                G_CIPHERS="$G_CIPHERS                           \
                    +RSA:+NULL:+SHA256                          \
                    "
            fi
```

When we run outcome analysis, we should treat `TLS-RSA-WITH-NULL-SHA256` as present, since we do expect to run it on the CI. So I think that `compat.sh --list-test-cases` should list it unconditionally (and in fact shouldn't try running `gnutls-cli` at all).

Actually, we should probably do this check differently: always include the cipher suite in the list, but skip it if GnuTLS doesn't support it for that protocol version. That should resolve the `--list-test-cases` output automatically.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
This is a faithful backport, but it's missing a weird edge case that's only in 2.x.

File: tests/compat.sh (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/compat.sh:287]: Tentative fix: https://github.com/gilles-peskine-arm/mbedtls/tree/backport-8586%2Bgnutls_null_sha256-list_always [CI job](https://mbedtls.trustedfirmware.org/job/mbedtls-restricted-release-new/2/)

Comment (mpg) [reply tests/compat.sh:287]: Thanks for noticing, analysing and suggesting a fix! Your fix looks good to me so I'm cherry-picking it.

**Comment by mpg** [comment]
Note: I've reviewed Gilles's commit and my cherry-picking means I'm approving it. So, I don't thing a 3rd reviewer is needed: Paul reviews everything, Gilles reviews my commit and I reviewed his.

File: tests/compat.sh (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/compat.sh:287]: Er, given the logs I think I messed something up, either causing tests not to be executed or spurious test cases to be logged or some naming convention mismatch.

Comment (mpg) [reply tests/compat.sh:287]: Aw. I'm investigating.

Comment (gilles-peskine-arm) [reply tests/compat.sh:287]: Ah, no, all of those were already not executed [before](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/PR-8956-head/3/execution/node/6696/log/?consoleFull), except for the old-version `+RSA:+NULL:+SHA256`, which are supposed to be executed on the CI so it's good that they're getting reported (and if they're getting executed but not being reported as such, that seems to be part of a larger preexisting issue).

**Comment by gilles-peskine-arm** [comment]
This looks good as a backport, but I think something else is missing, because unlike the [3.6 PR](https://github.com/Mbed-TLS/mbedtls/pull/8586), where the [logs](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/PR-8586-head/4/execution/node/8234/log/) show all compat tests executed, the logs [here](https://mbedtls.trustedfirmware.org/job/mbedtls-restricted-release-new/2/execution/node/6171/log/?consoleFull) report a very long list of test cases not executed.

**Review by gilles-peskine-arm — APPROVED** [review]
I'm approving this as a faithful backport, with no claim of completeness. The original PR finishes the job for 3.6, but we'll need to do more work in 2.28. That work can happen later.

**Review by gilles-peskine-arm — APPROVED** [review]
oops yes

**Review by paul-elliott-arm — APPROVED** [review]
LGTM
